### ISSUE TITLE
Move groundings to user code

### ISSUE THREAD

**Issue by wenkokke** [vehicle-lang/vehicle#502]
With respect to supporting different DLs, would it make sense to allow the user to provide the specification of the DL in Vehicle as a record (or just separate definitions), and ship a few well-known logics and operators with Vehicle in the standard library? When translating to Python, for each of the primitives, we could send a pair consisting of (1) the Vehicle function, and (2) the high-level name of the operator, if known, so we could use the more efficient abstractions in e.g. TensorFlow.

For instance, the user adds the following in their specification...
```agda
ProductRealLogic : Grounding
ProductRealLogic =
  { not     = \a -> 1 - a
  , and     = \a b -> a * b
  , or      = \a b -> a + b - (a * b)
  , implies = \a b -> 1 - a + (a * b)
  }
```
...and from command-line, they specify the grounding much like they would specify a property...

```
vehicle compile --target=loss --grounding=ProductRealLogic ...
```

And for a number of well-known logics, e.g., DL2, Godel, Lukasiewicz, Product, Yager, STL, and the others which have built-in support, we could implement the Grounding with primitives which, when sent to Python, communicate their high-level name in addition to their syntax tree, so they can be given special treatment in Python, if necessary.

(For reference, the word grounding as well as the product real logic, I've just taken as an example from the Logic Tensor Networks paper.)

**Comment by MatthewDaggitt** [comment]
Copying from Slack: 

Eventually yes it would make sense to do that. But I think it's really not a priority now. We have the infrastructure in Haskell for users to define their own grounding and I think it's going to be a pretty rare occurrence outside of that. Also we would need Vehicle to support records which would also be a major (but useful!) job.

**Comment by wenkokke** [comment]
With #513 we could potentially declare the logical connectives of vehicle itself as abstract, and provide different implementations for `Bool`, `LukasiewiczProp`, `GodelProp`, etc.

**Comment by wenkokke** [comment]
With #514 and #515 we could potentially get rid of all LossFunction code in Vehicle, and instead use `@compile` to defer interpretation of logical connectives to Python, where we would implement various interpretations of the standard library as instances of some protocol, i.e., an object that has a method for every builtin and every definition with an `@compile(JSON)` annotation.

**Comment by wenkokke** [comment]
With #513, #514, and #515, we could implement the operations of, e.g., the connectives for `GodelProp` in vehicle, mark it with `@compile(JSON)` to send it unnormalised, but allow Vehicle to add a context of declarations to the JSON file, just in case the Python backed doesn't have an implementation.

**Comment by wenkokke** [comment]
@MatthewDaggitt You had the concern that it would be odd for the GodelBool type to be different from the ClassicalBool type. What if the groundings lived in user code, but it was a Vehicle CLI flag that switched the interpretation of Bool?

**Comment by MatthewDaggitt** [comment]
My opinion is that users should not be really be writing groundings in `.vcl` files at all. The contract we have with the user is that the `.vcl` file is purely the specification. Whereas the groundings are part of the semantics.

**Comment by wenkokke** [comment]
Yeah, but it would let _us_ write groundings in vcl, and that's worth something.

The user is not really expected to write compile and noinline pragmas or abstract definitions either.

**Comment by MatthewDaggitt** [comment]
I don't think it's really worth the effort to let us write groundings in VCL. We can write them in Haskell just fine.

**Comment by MatthewDaggitt** [comment]
This was fixed in https://github.com/vehicle-lang/vehicle/pull/996
